components/src/dynamo/planner/monitoring/aic_power_optimizer.py

```
@@ -0,0 +1,635 @@
+
+        Strategy: maximise n_d first (decode is the throughput bottleneck for
+        OSL-heavy workloads), then use remaining budget for n_p.
+        Ties broken by fewer total GPUs, then lower cap power (§5.4).
+        """
+        # Maximum n_d given we must keep at least min_ep prefill replicas.
+        if budget is not None:
+            remaining_after_min_p = budget - min_ep * p_watts
+            if remaining_after_min_p <= 0:
+                logger.warning(
+                    "AIC optimizer: budget %dW cannot cover min_endpoint prefill "
+                    "replicas (%d × %dW); falling back to min_endpoint.",
```

Ruff (0.15.12)

[warning] String contains ambiguous `×` (MULTIPLICATION SIGN). Did you mean `x` (LATIN SMALL LETTER X)?

(RUF001)